json_serializable/lib/src/utils.dart

```diff
@@ -15,43 +15,61 @@ import 'package:source_gen/source_gen.dart';
 /// Dart code.
 // TODO: still need handle triple singe/double quotes within `value`
 String escapeDartString(String value) {
-  value = value.replaceAll('\\', r'\\');
+  var hasSingleQuote = false;
+  var hasDoubleQuote = false;
+  var hasDollar = false;
+  var canBeRaw = true;
+
   value = value.replaceAllMapped(_escapeRegExp, (match) {
+    var value = match[0];
+    if (value == "'") {
+      hasSingleQuote = true;
+      return value;
+    } else if (value == '"') {
+      hasDoubleQuote = true;
+      return value;
+    } else if (value == r'$') {
```

Comment:
You could treat `\` the same as `$` wrt. "canBeRaw", a string with content `"$\` can be raw as `r'"$\'`.

I can see that it doesn't work with the way this operation currently works - you can't decide later that the string can't be raw, because then you have to go back and fix previous `\` characters, so I guess this is a good enough trade-off since backslashes are rare.

The alternative is to scan the string once first to figure out what it contains, and then fix it up in one second operation.